:bug: `BufferedProcess.kill` now fully kills any child-process-eption that happens

acburdine opened the pull request:
closes #7252
- fixes :bug: where child processes spawned by spawned child processes on *nix systems were not getting killed correctly

acburdine commented:
[relevant article cred](http://krasimirtsonev.com/blog/article/Nodejs-managing-child-processes-starting-stopping-exec-spawn) for where this code comes from

thomasjo commented on src/buffered-process.coffee:
```diff
@@ -159,6 +160,27 @@ class BufferedProcess
+  killOnNix: ->
+    return unless @process?
+
+    self = this
```
:fire: This line can be removed

thomasjo commented on src/buffered-process.coffee:
```diff
@@ -159,6 +160,27 @@ class BufferedProcess
+
+    self = this
+    parentPid = @process.pid
+    psTree(parentPid, (err, children) ->
```
Use a fat arrow (`=>`) here to capture the scope of the function.

thomasjo commented on src/buffered-process.coffee:
```diff
@@ -159,6 +160,27 @@ class BufferedProcess
+        children.map((p) ->
+          return p.PID
+        )
+      ).forEach((tpid) ->
```
Rewrite the pid block as 
```coffee
pidsToKill = psTree parentPid, (err, children) -> [parentPid].concat(children.map (p) -> p.PID)
```
or equivalent.

thomasjo replied:
On second thought, considering we're invoking `@killProcess`, we don't want the parent PID here, so we should be able to change this code to
```coffee
pidsToKill = psTree parentPid, (err, children) -> children.map (p) -> p.PID
```

acburdine replied:
@thomasjo both of these seem to break the functionality introduced, as the array isn't returned by the (err, children) callback function

acburdine replied:
Also, I'm pretty sure psTree is asynchronous, meaning that it won't work to set it to something outside the callback. The forEach has to be handled inside the callback

thomasjo commented on src/buffered-process.coffee:
```diff
@@ -159,6 +160,27 @@ class BufferedProcess
+        try
+          process.kill(tpid)
+        catch ex
+      )
```
The `.forEach` block should be rewritten as
```coffee
for pid in pidsToKill
  try
    process.kill(pid)
```

thomasjo commented on src/buffered-process.coffee:
```diff
@@ -159,6 +160,27 @@ class BufferedProcess
+          process.kill(tpid)
+        catch ex
+      )
+      self.killProcess()
```
Make this `@killProcess()`.

acburdine commented:
@thomasjo It should be good to go now, checking to make sure changes still retain functionality

thomasjo commented on src/buffered-process.coffee:
```diff
@@ -159,6 +160,21 @@ class BufferedProcess
+
+    for pid in pidsToKill
+        try
+            process.kill(pid)
```
These two lines should be indented with 2 spaces, not 4.

thomasjo commented:
> Also, I'm pretty sure psTree is asynchronous, meaning that it won't work to set it to something outside the callback. The forEach has to be handled inside the callback

Aha, that changes the lay of the land slightly. I guess this should work then
```coffee
parentPid = @process.pid
psTree parentPid, (err, children) =>
  pidsToKill = children.map(child) -> child.PID

  for pid in pidsToKill
    try
      process.kill(pid)
  @killProcess()
```

acburdine commented:
Got it done, tested it so it should be good to go!

thomasjo commented on src/buffered-process.coffee:
```diff
@@ -159,6 +160,22 @@ class BufferedProcess
+        for pid in pidsToKill
+          try
+            process.kill(pid)
+        return
```
You can remove this `return` statement.

thomasjo commented on src/buffered-process.coffee:
```diff
@@ -159,6 +160,22 @@ class BufferedProcess
+    parentPid = @process.pid
+    psTree parentPid, (err, children) ->
+        pidsToKill = children.map (p) -> p.PID
+        for pid in pidsToKill
```
:art: The indentation here is wrong; it has to be 2 spaces, not 4.

thomasjo commented on src/buffered-process.coffee:
```diff
@@ -159,6 +160,22 @@ class BufferedProcess
+            process.kill(pid)
+        return
+
+    @killProcess()
```
Given the fact the `psTree` call is async, I think this invocation of `@killProcess` should be nested inside the callback handler (as seen in https://github.com/atom/atom/pull/9108#issuecomment-147311776). This is to guard against any possible side-effects of killing the primary process first. What do you think?

acburdine replied:
I would agree. I'll fix that

thomasjo replied:
Remember to change the outermost arrow to a fat arrow to capture the scope, otherwise `@killProcess()` is going to target the wrong scope. 

Thanks for sticking with me on this!

acburdine committed:
:bug: Kill now fully kills any child-process-eption that goes on

closes #7252
- fixes :bug: where child processes spawned by spawned child processes on *nix systems were not getting killed correctly

thomasjo commented:
Great, the code looks good to me now. Although I haven't tested this yet, which brings me to the next point; it would be great if you could add a test for this to ensure we're protected against regressions in functionality.

Maybe we should add a separate `describe` block for `"on Linux and OS X"` (or some such name) along the lines of the `"on Windows"` block. And given the added functionality is dependent upon actually running on a *NIX system, the block should not run on Windows, so simply `return` if the platform is set to `win32`.

/cc @atom/feedback

acburdine commented:
@thomasjo to that effect, shouldn't the killOnWindows function also have a check to make sure it is windows?

acburdine commented:
also, I'm not 100% sure how to test for the behavior. I know the windows spec block tests to make sure the process is wrapped in a cmd.exe wrapper, but I'm not sure how to run a command that spawns another child process and keeps it running

thomasjo commented:
> to that effect, shouldn't the killOnWindows function also have a check to make sure it is windows?

I was specifically talking about specs in that context, so we are not changing e.g. `killOnWindows` or `killOnNix` to check for platforms. What we need to do though, is to ensure we don't try to run the specs covering `killOnNix` when we're on a Windows platform.

thomasjo commented:
If you don't feel up to the challenge with regards to adding specs for this — I realize testing this is not trivial — perhaps one of folks from @atom/non-github-maintainers (or someone else) would like to assist. If nobody hops on this, I'll try to find some time by the end of the week.

acburdine commented:
Makes sense. I'll try and get this finished up over the next couple of days.

acburdine commented:
@thomasjo I've racked my brain trying to figure out a way to test this out. I can't think of any system commands that would spawn their own child processes (except maybe a fork bomb, but that's probably not a good idea :wink:). I'm not sure exactly where to go from here.

thedaniel commented:
> I can't think of any system commands that would spawn their own child processes

probably [foreman](https://github.com/ddollar/foreman), or perhaps a web server that does concurrency via child processes?

acburdine commented:
@thedaniel yeah that would work, the only issue is that the test suite would have to install foreman just for doing testing. There are a lot of npm packages as well that could work to test this behavior out, but to add another package solely for testing this one behavior seems like overkill to me.

I'm willing to write a test using an npm library to generate the child processes, but only if it's deemed worth adding.

thedaniel commented:
ah, sorry, i just skimmed this and didn't get the full context.

1Uno commented:
Does this really work? This is just terminology ain't it. No killing
